JDK-8308659: Use CSS scroll-margin instead of flexbox layout in API documentation

hns opened the pull request:
A few years ago we switched to [Flexbox Layout](https://developer.mozilla.org/en-US/docs/Glossary/Flexbox) to implement the sticky header in the API docs generated by `javadoc`. This solved the problem of anchor link targets [not being positioned correctly](https://bugs.openjdk.org/browse/JDK-8223378), but it also introduced a few new ones:

 - It required a workaround to get browser history to work (JDK-8249133, JDK-8250779, 8286832)
 - It changed certain aspects of scrolling behavior in the browser (JDK-8301080)
 - It changed the way some CSS properties are interpreted (JDK-8315800)

The reason for most of these problems is that the layout paradigm used by Flexbox is very different from traditional layout of HTML pages. The `scroll-margin-*` CSS properties introduced by the [CSS Scroll Snap Module](https://www.w3.org/TR/css-scroll-snap-1/) provide a simpler and less intrusive way to implement link target offsets in combination with sticky elements implemented using [`position: sticky`](https://developer.mozilla.org/en-US/docs/Web/CSS/position). However, although it is implemented [in all supported browsers](https://caniuse.com/?search=scroll-margin), it comes with its own challanges and quirks, which are explained below.

The first problem to overcome was that `position: sticky` is more fragile on mobile browsers than Flexbox. If some part of the page content overflows the allotted horizontal space, the whole page can be zoomed out to view the whole content. This causes the header to become unsticky and the link target offsets to become erroneous. It was therefore necessary to make sure page content never overflows its allotted horizontal space, either by adding line break opportunities, or by making all or part of the page horizontally scrollable when its content overflows. Line break opportunities are difficult to add especially in  preformatted code, so I opted for making the parts of the page most likely containing long lines of code scrollable. 

The next problem was that enabling horizontal scrolling on an element or one of its containing elements breaks the `scroll-margin-top` property in Chrome due to a browser quirk (both desktop and mobile versions). This means that the scrolling must occur in a child element rather than the sections or other elements serving as link targets. 

When enabling horizontal scrolling on the contents of sections containing user-provided content, another problem is that it disables [Margin Collapse](https://www.joshwcomeau.com/css/rules-of-margin-collapse/) which regulates margins between adjacent and contained boxes. Since these sections can contain a mix of HTML containers, it is almost impossible to preserve the margins between elements while making them scrollable. 

The solution to these problems was to add a new `<div class="horizontal-scroll">` to the sections containing user-provided content. Adding a single scrollable element per section allowed the margins to be controlled in a relatively safe way, although some CSS is introduced to limit the margin of certain trailing block elements in certain sections. 

The `scroll-margin-top` property is set to the normal height of the sticky header in `stylesheet.css`, with a line added to `search.js` to set the property dynamically to the actual header height to accommodate for the pre-release notification banner. In a previous version of this branch, the pre-release banner was excluded from the sticky header, but it was made sticky again in the final version to make the output more comparable. 

The generated documentation should render the same on supported browsers pixel by pixel. Before/after snapshots of the JDK docs are available here:

Before: https://cr.openjdk.org/~hannesw/8308659/api.00/
After: https://cr.openjdk.org/~hannesw/8308659/api.01/

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308659](https://bugs.openjdk.org/browse/JDK-8308659): Use CSS scroll-margin instead of flexbox layout in API documentation (**Enhancement** - P4)


### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to [fcfcc5fe](https://git.openjdk.org/jdk/pull/15969/files/fcfcc5fea7c3c8382f3a62d4ce3418cd22cd9493)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15969/head:pull/15969` \
`$ git checkout pull/15969`

Update a local copy of the PR: \
`$ git checkout pull/15969` \
`$ git pull https://git.openjdk.org/jdk.git pull/15969/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15969`

View PR using the GUI difftool: \
`$ git pr show -t 15969`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15969.diff">https://git.openjdk.org/jdk/pull/15969.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15969#issuecomment-1749125457)

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ConstructorWriter.java:
```diff
@@ -106,13 +106,13 @@ protected void buildConstructorDoc(Content target) {
-                buildConstructorComments(constructorContent);
-                buildTagInfo(constructorContent);
-
+                Content scrollBox = HtmlTree.DIV(HtmlStyle.horizontalScroll);
```
Stylistically, it's a shame to change the name of the variable in this and similar situations when all you are changing is its initial value. In the lines that follow, its only important that it is some kind of container, not specifically a `scrollBox`.  This is the variable name equivalent of whether to write
```
Map m = new HashMap();
```
or 
```
HashMap m = new HashMap();
```

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/script.js:
```diff
@@ -227,27 +227,3 @@ function switchCopyLabel(button, span) {
         }, 100);
     }, 1900);
 }
-// Workaround for scroll position not being included in browser history (8249133)
-document.addEventListener("DOMContentLoaded", function(e) {
-    var contentDiv = document.querySelector("div.flex-content");
-    window.addEventListener("popstate", function(e) {
-        if (e.state !== null) {
-            contentDiv.scrollTop = e.state;
-        }
-    });
-    window.addEventListener("hashchange", function(e) {
-        history.replaceState(contentDiv.scrollTop, document.title);
-    });
-    var timeoutId;
-    contentDiv.addEventListener("scroll", function(e) {
-        if (timeoutId) {
-            clearTimeout(timeoutId);
-        }
-        timeoutId = setTimeout(function() {
-            history.replaceState(contentDiv.scrollTop, document.title);
-        }, 100);
-    });
-    if (!location.hash) {
-        history.replaceState(contentDiv.scrollTop, document.title);
-    }
-});
```
always nice to see workarounds removed!

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlStyle.java:
```diff
@@ -964,6 +941,11 @@ public enum HtmlStyle {
+    /**
+     * The class of a {@code div} element that allows its horizontal overflow to be scrolled.
+     */
+    horizontalScroll,
```
quibble: scroll? or scrollable?

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/MethodWriter.java:
```diff
@@ -105,13 +105,13 @@ protected void buildMethodDoc(Content detailsList) {
             for (Element method : methods) {
                 currentMethod = (ExecutableElement)method;
```
As a separate RFE, _it would be nice_ if `getVisibleMembers` took an optional `Class<T>` argument that is used to cast the result of the method to `List<T extends Element>` instead of just `Element`.  This would enable to eliminate downstream casts, like here on line 106, to get the subtype.

See a similar technique in `Utils` for accessing sublists of a list of `DocTree` nodes,

jonathan-gibbons commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/MethodWriter.java:
```diff
@@ -105,13 +105,13 @@ protected void buildMethodDoc(Content detailsList) {
-                buildMethodComments(methodContent);
-                buildTagInfo(methodContent);
-
+                Content div = HtmlTree.DIV(HtmlStyle.horizontalScroll);
+                buildSignature(div);
+                buildDeprecationInfo(div);
+                buildPreviewInfo(div);
+                buildMethodComments(div);
+                buildTagInfo(div);
+                methodContent.add(div);
```
Purely as a matter of style, I like the style of building "bottom up", so first build all those items into a `ContentBuilder`, then make a `div` with `HtmlTree.DIV(HtmlStyle.horizontalScroll, contentBuilder)`  then add the `div` into the `methodContent`, as on line 114.

hns commented:
I have updated this PR with a [few additional fixes](https://github.com/openjdk/jdk/pull/15969/commits/aad899048bdc5cd70424ab73bdcb1c2efa0eda4f):

 - The callback to update the scroll-margin to accomodate for the draft header has moved from `search.js` to `script.js` which is always available (`search.js` is dependent on the index).
 - The callback is implemented in pure JavaScript instead of relying on jQuery, and uses the `DOMContentLoaded` event instead of the `window` `load` event, which avoids flickering on Chrome and Safari (still a little bit of flickering in Firefox under some conditions).
 - I removed the change to make tab buttons scrollable instead of adding line breaks as the `:has()` CSS pseudo-class is not supported on Firefox. This change was unrelated to the issue at hand.

Generated documentation can be viewed here: https://cr.openjdk.org/~hannesw/8308659/api.03/

hns commented:
/integrate
